runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -807,6 +902,145 @@ JITServerAOTCacheMap::buildCacheFileName(const std::string &cacheDir, const std:
+            }
+         else // Header is compatible, check the number of methods
+            {
+            if (cache->getNumCachedMethods() > header._numMethods)
+               {
+               // We have better data than the existing snaphot, so overwrite it
+               dosave = true;
+               // Ideally, if we know that our cache has fewer entries than the existing snapshot,
+               // we shouldn't try again to save it (unless meanwhile we store more methods in ours)
+               // However, somebody else may overwrite the cache with an incompatible one
+               // At the very least we have to write the time when the attempt took place
+               }
+            }
+         }
+      else // Read error
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+           TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: Cannot read header from exiting cache file %s", cacheFileName.c_str());
+         // We should still proceed with the save operation, because the current cache is not usable
+         doSave = true;
+         }
+      */
+      fclose(cacheFile);
+      cacheFile = NULL;
+      }
+   else // Cannot open file
+      {
+      // If we cannot open the file for reading, the file might not exist
+      // and this is the first attempt to save a cache. It's also possible
+      // that the file exists, but we cannot read it. It's unlikely that
+      // the file has write permissions, but not read permissions
+      doSave = true;
+      }
+
+   if (doSave)
+      {
+      // Create a temporary file based on the UID of this server and the cache name
+      std::string tempFileName = buildCacheFileName(compInfo->getPersistentInfo()->getJITServerAOTCacheDir(),
+                                                    std::to_string(compInfo->getPersistentInfo()->getServerUID()) + cacheName + ".tmp");
+      FILE *newCacheFile = fopen(tempFileName.c_str(), "wb");
+      if (newCacheFile)
+         {
+         if (cache->writeCache(newCacheFile)) // TODO: can the cache be deleted since we last checked?
+            {
+            fclose(newCacheFile);
+            // Rename the file to the final name
+            if (0 == rename(tempFileName.c_str(), cacheFileName.c_str()))
+               {
+               success = true;
+               if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+                  TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: Saved cache %s to file %s", cacheName.c_str(), cacheFileName.c_str());
+               }
+            else // Renaming failed
+               {
+               if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+                  TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: Failed to rename temporary cache file %s to %s", tempFileName.c_str(), cacheName.c_str());
+               remove(tempFileName.c_str());
+               cache->excludeCacheFromSavingToFile();
```

Comment:
We should mention in the "documentation" comment for this method (or elsewhere) that we won't try writing again in case of file creation or rename failure (which makes sense because either there are permission issues or we ran out of storage space).

Reply:
Done